Upgrade shadow-cljs and ClojureScript

ilmotta opened the pull request:
### Summary

This PR upgrades shadow-cljs from `2.11.16` (released on 2021-02-03) to `2.19.6` (released on 2022-07-14), so ~1.5 years worth of upgrades. By upgrading shadow-cljs we can finally use the latest major Clojure version `1.11`.

### Why upgrade shadow-cljs?

- Shadow-cljs controls the ClojureScript version we can use. In order to use the latest major Clojure version (`1.11.x`) we need to upgrade shadow-cljs.

- Shadow-cljs releases new versions very frequently, and if you take a look at its [changelog](https://github.com/thheller/shadow-cljs/blob/master/CHANGELOG.md), you'll see it had tons and tons of bug fixes over the years. I hope some of them help improve the DX for certain contributors who recently reported issues with shadow-cljs. It might make things worse though, we never know.

- Clojure 1.11 brings new features, bug fixes and even performance improvements (although I think the performance mostly impacts Clojure on the JVM). See the changelog https://github.com/clojure/clojure/blob/master/changes.md#changes-to-clojure-in-version-1110

Things that can be beneficial to us, or are interesting nonetheless:

- New `:as-alias` to be used in `require`, which is like `:as` but does not require the namespace to load. This means namespaced keywords using `:as-alias` can't cause circular dependency errors. This feature would very useful if we used namespaced keywords, but we don't, so... https://github.com/clojure/clojure/blob/master/changes.md#22-as-alias-in-require
- New macros `run-test` and `run-test-var` to run single test with fixtures and report.
- New `iteration` function, useful for processing paginated data. https://www.abhinavomprakash.com/posts/clojure-iteration/
- New `update-keys` function: applies a function to every key in a map.
- New `update-vals` function: applies a function to every value in a map.

Examples for `update-vals` and `update-keys`. They should perform better than the common `reduce-kv` approach since they use a transient data structure.

```clojure
(let [m {:a 1 :b 2}]
  (update-vals m inc)) ; => {:a 2, :b 3}

(let [m {:a 1 :b 2}]
  (update-keys m name)) ; => {"a" 1, "b" 2}
```

### Why change namespaces within `__tests__` directories?

Any namespace with the word `--tests--` throws an error, like the one below. I didn't bother investigating why, so I changed the guidelines to reflect the new convention. It's probably related to the double dashes in the name :man_shrugging:

```
Namespace quo2.components.dividers.--tests--.divider-label-component-spec has a segment starting with an invalid JavaScript identifier at line 1
```

### Why can't we upgrade to the most recent shadow-cljs version?

We can't go to versions newer than `2.19.6` because we're stuck with the very old OpenJDK 8, and some new code from shadow-cljs `2.19.7+` requires at least OpenJDK 11.

### Review notes

Please, review this PR by running the app as you usually do, especially if you have access to macOS, since I don't have one at hand.

### Steps to test

Things I tested:

- Running unit tests with or without the REPL.
- Running component tests.
- Re-frisk (since I had to hard code a new version to remove a warning from the `encore` lib).
- Released APK on an Android device.
- Smoke tests in the app (clicking around).

status: ready

ilmotta commented on src/status_im2/common/bottom_sheet/view.cljs:
```diff
@@ -16,14 +16,11 @@
-            options)
-   callback))
+  [value]
+  (reanimated/with-spring value
```
FYI: The newer ClojureScript version throws a runtime error when the original function signature is used. I just removed the extra arguments, since they were never used in the first place.

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 54d0a894 | [#1](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-15417/1/) | 2023-03-21 00:18:41 | ~6 min | `tests` | [:page_facing_up:`log`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-230321-001205-54d0a8-pr15417-tests.log) |

siddarthkay reviewed (commented):
great work! 
Hopefully when the `react-native` upgrade gets merged in we have java upgraded to `openjdk11` in that branch. 
That should further unblock you to experiment with latest `shadow-cljs`.

flexsurfer commented on shadow-cljs.edn:
```diff
@@ -49,7 +55,11 @@
+                            ;; $jscomp (Clojure Compiler) won't be defined.
+                            ;; Going for :es6 is a must for upgrading to
+                            ;; ShadowCLJS 2.15.13+.
+                            :output-feature-set :es6
```
https://github.com/thheller/shadow-cljs/issues/1080#issuecomment-1412641778

flexsurfer replied:
this won't work with hermes engine

ilmotta replied:
Thanks @flexsurfer I'll check again, that's quite unfortunate. I have Hermes disabled in my environment, that's probably why it worked.

flexsurfer replied:
yeah, in latest RN its enabled by default on ios

jakubgs reviewed (approved):
I think you might have forgotten to run `make nix-update-clojure`.

J-Son89 reviewed (approved):
Nice one @ilmotta!!

vkjr commented:
Cool! I'm glad we'll have `update-vals` 😍

ilmotta commented:
> > We can't go to versions newer than 2.19.6 because we're stuck with the very old OpenJDK 8
> 
> @ilmotta btw why can't we upgrade OpenJDK?

@OmarBasem, I don't know the answer to this question because I've never tried to upgrade. @siddarthkay our RN version is tied to Java 8, so if we upgrade RN to let's say 0.70, then we might be able to go to Java 11.

But right now I'm having trouble upgrading shadow-cljs due to the need to support Hermes, but in order to support Hermes we need to compile CLJS to the ES5 target, but the ES5 target doesn't work with newer versions of shadow-cljs (at least in `status-mobile` repo). It's a can of worms...

Unfortunately I might just close this PR if I can't find a solution until tomorrow :/

OmarBasem commented:
> Unfortunately I might just close this PR if I can't find a solution until tomorrow :/

@siddarthkay are we expecting the RN upgrade anytime soon :)

siddarthkay commented:
@OmarBasem : can't say when, some serious Yak Shaving going on here : https://github.com/status-im/status-mobile/pull/15203
All I can say is that the RN Upgrade has the attention it deserves and we will get there soon :)

ilmotta commented:
> @OmarBasem : can't say when, some serious Yak Shaving going on here : #15203 All I can say is that the RN Upgrade has the attention it deserves and we will get there soon :)

Yes, there's some seriously good work going on the upgrade to RN! :100: :heart: 

@OmarBasem I think it's fine if we stay on Java 8 until all the dust settles because we don't need to go to the very latest of shadow-cljs & ClojureScript. But we'll eventually get to OpenJDK 11 and beyond, I'm sure!

ulisesmac commented:
> But right now I'm having trouble upgrading shadow-cljs due to the need to support Hermes, but in order to support Hermes we need to compile CLJS to the ES5 target, but the ES5 target doesn't work with newer versions of shadow-cljs (at least in `status-mobile` repo). It's a can of worms...
> 

Hey @ilmotta 
What's the problem when compiling to ES5?

Some months ago(~3), I created a small RN app using a recent version of shadow-cljs, Hermes enabled and compiled to ES5. It ran without any problems.

Also, Hermes can be disabled so that we can compile to ES6.

ilmotta commented:
> > But right now I'm having trouble upgrading shadow-cljs due to the need to support Hermes, but in order to support Hermes we need to compile CLJS to the ES5 target, but the ES5 target doesn't work with newer versions of shadow-cljs (at least in `status-mobile` repo). It's a can of worms...
> 
> Hey @ilmotta What's the problem when compiling to ES5?
> 
> Some months ago(~3), I created a small RN app using a recent version of shadow-cljs, Hermes enabled and compiled to ES5. It ran without any problems.
> 
> Also, Hermes can be disabled so that we can compile to ES6.

@ulisesmac, it simply doesn't work, on my machine at least, but since we're all running through Nix, it probably won't work for you too. You can try on this branch to change the shadow-cljs setting to ES5 and it'll fail saying the G. Closure Compiler is not available.

> Also, Hermes can be disabled so that we can compile to ES6.

Hermes is disabled on my machine too, but as @flexsurfer said, it's the default engine when running on macOS. If we could just say that we won't ever use Hermes then I would have been able to merge this PR already.

@flexsurfer, do we need to support Hermes at all? Hermes is already problematic for some folks, including me. When it's enabled it causes crashes on my GNU/Linux environment. What's your opinion as a macOS user?

flexsurfer commented:
so, we're trying to use a tool, which allows you to set a property but it doesn't work, right, so we need to file an issue in shadow-cljs repo i guess? hermes is used in release builds, and it should be faster, and its default engine in RN, so im not sure disabling it is a good idea

ilmotta commented:
> so, we're trying to use a tool, which allows you to set a property but it doesn't work, right, so we need to file an issue in shadow-cljs repo i guess? hermes is used in release builds, and it should be faster, and its default engine in RN, so im not sure disabling it is a good idea

I agree, that's the way forward. Thanks for sharing.

ilmotta commented:
@ulisesmac, running the command @jakubgs kindly [reminded me](https://github.com/status-im/status-mobile/pull/15417#pullrequestreview-1349890686) did the trick, so it's working with/without Hermes. Thanks for reviewing the PR. Could you please check if this branch works on your Linux distro?

@J-Son89, @OmarBasem, @flexsurfer, could one of you please verify if the branch works with Hermes enabled. I know you guys use macOS :) I rebased against `develop` now.

flexsurfer reviewed (approved):
thank you @ilmotta

ulisesmac commented:
> @ulisesmac, running the command @jakubgs kindly [reminded me](https://github.com/status-im/status-mobile/pull/15417#pullrequestreview-1349890686) did the trick, so it's working with/without Hermes. Thanks for reviewing the PR. Could you please check if this branch works on your Linux distro?
> 
> @J-Son89, @OmarBasem, @flexsurfer, could one of you please verify if the branch works with Hermes enabled. I know you guys use macOS :) I rebased against `develop` now. Everything should just work ;)

I'm using Fedora, I'll try it today, sorry for the waiting.

flexsurfer commented:
here https://github.com/status-im/status-mobile/issues/14776 i used latest versions of RN and shadow and it worked just fine

ilmotta commented:
> here #14776 i used latest versions of RN and shadow and it worked just fine

That's great news @flexsurfer! So as soon as the RN upgrade lands in `develop` we'll be good to go!

OmarBasem commented:
Hey @ilmotta, any chance to try the shadow-cljs upgrade again as we are using Java 11 now

ilmotta commented:
Hi @OmarBasem. I can imagine you're still suffering from shadow-cljs issues. Let me give this issue another good try today and report back the results.

ilmotta committed:
Upgrade to Shadow CLJS 2.11.26

ilmotta committed:
Upgrade Shadow CLJS to 2.12.0 and oops to 0.7.2

ilmotta committed:
Upgrade Shadow CLJS to 2.15.11

ilmotta committed:
Upgrade Shadow CLJS to 2.15.13 (broken with Hermes)

ilmotta committed:
Enable feature set ES6 (still broken)

ilmotta committed:
Upgrade Shadow CLJS to 2.17.0

https://ask.clojure.org/index.php/11267/clojurescript-handle-functions-accept-keywords-differently

ilmotta committed:
Upgrade to Shadow CLJS 2.19.3 and remove warning from taoensso.encore

    ------ WARNING #1 - :redef -----------------------------------------------------
     Resource: taoensso/encore.cljs:947:1
    --------------------------------------------------------------------------------
     944 |                                                                                         )
     945 |
     946 | (defn pow [n exp] (Math/pow n exp))
     947 | (defn abs [n]     (if (neg? n) (- n) n)) ; #+clj (Math/abs n) reflects
    -------^------------------------------------------------------------------------
     abs already refers to: cljs.core/abs being replaced by: taoensso.encore/abs
    --------------------------------------------------------------------------------
     948 | (defn round* ; round
     949 |   ([             n] (round* :round nil n))
     950 |   ([type         n] (round* type   nil n))
     951 |   ([type nplaces n]
    --------------------------------------------------------------------------------

ilmotta committed:
Upgrade to Shadow CLJS 2.19.6 and fix compilation errors

ilmotta committed:
Upgrade Shadow CLJS to 2.19.7

ilmotta committed:
Upgrade Shadow CLJS to 2.25.0 (2023-07-26)

ilmotta committed:
Fix: rely on shadow logic to set the output feature set

ilmotta committed:
Go back to original (develop) settings for the feature set

ilmotta commented:
@OmarBasem, whenever you can, please test this branch with Hermes enabled and if possible, check if you can do a local release to iOS (because I can't yet test this). I've reopened the PR because I managed to upgrade to the latest Shadow CLJS and as expected, the accompanying latest ClojureScript. If you report no issues on your side I'll again request reviews from the other devs.

OmarBasem commented:
Much appreciated @ilmotta. I will try out the branch and update you 👍

OmarBasem commented:
Hi @ilmotta, I have tried out the branch and it seems to be working well. 👍 

Building on Android with Hermes enabled works fine.

But I wasn't able to test local iOS release builds. For me building local ios releases fails -- even on develop, so I guess it is not something specific to this PR. I see that the iOS build in CI has succeeded.

ilmotta commented:
Hey folks, I re-requested your reviews and resurrected this PR (thanks @OmarBasem) because I could upgrade Shadow CLJS and ClojureScript to their latest versions after the amazing work done in the mobile team to upgrade RN 🚀 

I haven't seen any issues in my tests with Android with and without Hermes and in dev or release builds.

**If you can, please report back if you tested on iOS and everything is good**. That way we can quickly merge this PR.

Thank you! 💟

siddarthkay reviewed (approved):
Great work @ilmotta!
I tested the PR build on my iPhone 11 and it works fine!

OmarBasem reviewed (approved):
🚀

flexsurfer reviewed (approved):
nice!

churik commented:
Thanks for your amazing work!
e2e failures are known but not related to PR.
Tested basic flows, LGTM!

**Devices:**
- IPhone 14 Pro (IOS 16.5.1)
- Google Pixel 7A (Android 13)

ilmotta commented:
Oh @churik, thank you so much, I was going to move the PR to the e2e column but forgot in the myriad of things going on today. Great news, I want this PR to be merged real bad ;)
